cloudinit/sources/DataSourceWSL.py

```diff
@@ -316,11 +334,35 @@ def _get_data(self) -> bool:
         if not any([user_data, agent_data]):
             return False
 
-        # If we cannot reliably model data files as dicts, then we cannot merge
-        # ourselves, so we can pass the data in ascending order as a list for
-        # cloud-init to handle internally
-        if isinstance(agent_data, bytes) or isinstance(user_data, bytes):
-            self.userdata_raw = cast(Any, [user_data, agent_data])
+        # If only user_data was found
+        if agent_data is None:
+            assert user_data is not None
+            self.userdata_raw = user_data.raw()
+            return True
+
+        # If only agent data was found
+        if user_data is None:
+            assert agent_data is not None
+            self.userdata_raw = agent_data.raw()
+            return True
+
+        # If both are found but we cannot reliably model both data files as
+        # cloud-config dicts, then we cannot merge them ourselves, so we should
+        # pass the data as if the user had written an include file
+        # for cloud-init to handle internally. We explicitely prioritize the
+        # agent data, to ensure cloud-init would handle it even in the presence
+        # of syntax errors in user data (agent data is autogenerated).
+        # It's possible that the effects caused by the user data would override
+        # the agent data, but that's the user's ultimately responsibility.
+        # The alternative of writing the user data first would make it possible
+        # for the agent data to be skipped in the presence of syntax errors in
+        # user data.
+        assert user_data is not None and agent_data is not None
+        if not agent_data.is_dict() or not user_data.is_dict():
```

Comment:
Let's change this to a simple ConfigData.is_cloud_config() check to better represent what we are looking for here. If not cloud_config, and non-empty, use `#include` directive.